readthedocs/invitations/models.py

```diff
@@ -0,0 +1,209 @@
+                "from_user": from_user,
+            },
+        )
+        if created:
+            invitation.send()
+        return invitation, created
+
+    def for_object(self, obj):
+        return self._for_object(obj=obj, queryset=self.all())
+
+    @staticmethod
+    def _for_object(obj, queryset):
+        return queryset.filter(
+            object_id=obj.pk,
+            content_type=ContentType.objects.get_for_model(obj),
+        )
+
+
+class Invitation(TimeStampedModel):
+
+    """
+    Invitation model.
+
+    An invitation can be attached to an existing user or to an email.
+    """
+
+    # Generic foreign key.
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    object = GenericForeignKey("content_type", "object_id")
+
+    # Normal fields.
+    from_user = models.ForeignKey(
+        User,
+        verbose_name=_("From user"),
+        on_delete=models.CASCADE,
+        related_name="invitations_sent",
+    )
+    to_user = models.ForeignKey(
+        User,
+        verbose_name=_("To user"),
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="invitations_received",
+    )
+    to_email = models.EmailField(_("E-mail"), null=True, blank=True)
+    token = models.CharField(unique=True, max_length=32)
+
+    objects = InvitationQueryset.as_manager()
+
+    class Meta:
+        indexes = [
+            models.Index(fields=["content_type", "object_id"]),
+        ]
+        unique_together = [
+            ("to_user", "content_type", "object_id"),
+            ("to_email", "content_type", "object_id"),
+        ]
+
+    @property
+    def username(self):
+        if self.to_user:
+            return self.to_user.username
+        return self.to_email
+
+    @property
+    def expired(self):
+        return timezone.now() > self.expiration_date
+
+    @property
+    def expiration_date(self):
+        return self.created + timezone.timedelta(
+            days=settings.RTD_INVITATIONS_EXPIRATION_DAYS
+        )
+
+    @cached_property
+    def backend(self):
+        return get_backend(self)
+
+    def save(self, *args, **kwargs):
+        if not self.token:
+            self.token = self.generate_token()
+        super().save(*args, **kwargs)
+
+    @staticmethod
+    def generate_token():
+        return get_random_string(32)
+
+    def redeem(self, user=None):
+        """
+        Redeem invitation.
+
+        `user` will be used only if the invitation is attached
+        to an email, otherwise `to_user` user will be used.
+        """
+        if self.expired:
+            return False
+        if self.to_user:
+            user = self.to_user
+        log.info(
+            "Redeeming invitation",
+            invitation_pk=self.pk,
+            for_user=user.username,
+            object_type=self.object_type,
+            object_name=self.object_name,
+            object_pk=self.object.pk,
+        )
```

Comment:
I like that we have traceability of who invited who. For instance, according to GDPR, you should provide a log of individuals with access to certain classes of data at a given time. And since we don't know what exact data is part of a documentation project, that might be relevant.

The log stores almost exactly what the database stores, except it adds a very convenient timestamp! Which is great, we can have this information stored two places, since the log will be more of a true "auditable" trail.

But I think we should also store the date of when an invitation was redeemed, so almost the same as the log's timestamp...

Is it possible that `redeem()` sets a field `self.redeemed = timezone.now()`?

This way, we have this fundamental bit of data logged and available in the database for future needs.